text/3389-manifest-lint.md

```diff
@@ -0,0 +1,250 @@
+
+Instead of traditional workspace inheritance where there is a single value to
+inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
+which defines presets and the user could do `lints.<preset> = true`.  The user
+could then name them as they wish to avoid collision with rustc lints.
+
+Instead of the `[lints]` table being `lint = "level"`, we could organize
+it around `level = ["lint", ...]` like some other linters do (like
+[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
+logically organizing lints, highlighting what changed in diffs, and for
+possibly adding lint-specific configuration in the future.
+
+# Prior art
+[prior-art]: #prior-art
+
+Python
+- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
+  - Format is `level = [lint, ...]`
+- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
+  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
+- [ruff](https://beta.ruff.rs/docs/configuration/)
+  - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)
+
+Javascript
+- [eslint](https://eslint.org/docs/latest/use/configure/rules)
+  - Format is `lint = level` / `lint = [ level, additional config ]`
+
+Go
+- [golangci-lint](https://golangci-lint.run/usage/configuration/)
+  - Format is `level = [lint, ...]`
+
+Ruby
+- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
+  - Format is `Lint: Enabled: true`
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should we hand rustdoc lint levels or, in the future, cargo lint levels?
+The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
+but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
+to get into
+[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).
+
+Should we only apply/fingerprint lints for the appropriate tool?  For example,
+we would not include and fingerprint `clippy::` lints when running builds,
+allowing them to change without forcing a rebuild.  We likely already need to
+be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
+this isn't much more of a step.
+
+How do we allow controling precedence between lints and lint groups?  We are
+using a TOML table with the keys as lint names which does not allow controlling
+ordering.  Even if we switched to `level = [lint, ...]`, you get a hard coded
+precedence between levels that the user can't control.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Configurable lints
+
+We can extend basic lint syntax:
+```toml
+[lints]
+cyclomatic_complexity = "allow"
+```
+to support configuration, whether for cargo or the lint tool:
+```toml
+[lints]
+cyclomatic_complexity = { level = "allow", rust-version = "1.23.0", threshold = 30 }
```

Comment:
This syntax is inverted compared to `#[allow(lint1, lint2)]`, which surprised me.

I find the form below clearer and it encourages clearly separating lints by level

```toml
[lints]
allow = [
    "unsafe",
    { name = "cyclomatic_complexity", rust-version = "1.23.0", threshold = 30 },
]
warn = [
    # ...
]
```

It's probably closer to how `(dev-)dependencies` work too:  we don't specify whether the dependency is dev, build or normal in `clap = { ... }`, we specify it outside

Ultimately I think both are fine, I just prefer the above form with lists.

The downside is that changing a lint from one level to another is now a bigger diff and less "local"

Reply:
This was covered down in the Alternatives section

> Instead of the `[lints]` table being `lint = "level"`, we could organize
it around `level = ["lint", ...]` like some other linters do (like
[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
logically organizing lints, highlighting what changed in diffs, and for
possibly adding lint-specific configuration in the future.

Yes, technically, we can put inline tables into inline arrays, like your example, but it is getting a bit messier.

Reply:
I agree that this feels way more natural. Also with allowing `{ ... }` instead of just lint names in the list, we also get configurability here. However, this would necessitate to define a lint level in order to configure lints, which is probably unnecessary in most cases.

Reply:
> This was covered down in the Alternatives section

Ah sorry, missed it. 

> I forgot to include in that quoted block that you also have to define the semantics for when a lint appears in multiple levels.

An error from cargo should be enough ?

---

> works better for logically organizing lints, highlighting what changed in diffs, and for possibly adding lint-specific configuration in the future.

I don't think that's true apart from diffs

In all projects where I have a `[target.*.rustflags]`, the lints are separated by allow/warn/deny and then sorted if necessary. In `clippy.toml` the configuration are sorted per-lint but there is no notion of level there.

> lint-specific configuration in the future.

How does the following work in toml ? Does one overrides the other ?

```toml
[lints.allow."unsafe"]
config1 = true

[lints]
allow = ["lint-2", "lint-3"]
```

Reply:
> How does the following work in toml ? Does one overrides the other ?

Assigning an array to an implicit table is a hard error.

Reply:
ah, there goes my hope, then you solution is probably best